cms/models.py

```diff
@@ -169,6 +170,20 @@ class ProgramIndexPage(CourseObjectIndexPage):
 
     def get_child_by_readable_id(self, readable_id):
         """Fetch a child page by the related Program's readable_id value"""
+        queryset = ProgramRun.objects.annotate(
+            program_run=Concat(
+                "program__readable_id",
+                models.Value(ENROLLABLE_ITEM_ID_SEPARATOR),
+                "run_tag",
+            )
+        )
```

Comment:
We only need to check for an existing program run if the readable id matches a program run id pattern. We run a check like this in `ecommerce.api.get_product_from_text_id`. If we check for the program run pattern first, we can avoid running a database query when we don't need to. This implementation should be changed to look something like this:

```python
program_run_id_match = re.match(PROGRAM_RUN_ID_PATTERN, text_id)
# This text id matches the pattern of a program text id with a program run attached
if program_run_id_match:
  match_dict = program_run_id_match.groupdict()
  if ProgramRun.objects.filter(program__readable_id=match_dict["text_id_base"], run_tag=match_dict["run_tag"]).exists():
    # If the given readable_id matches a ProgramRun, remove the run tag from the 
    # readable_id (example: `program-v1:my+program+R1` -> `program-v1:my+program`)
    readable_id = match_dict["text_id_base"]
self.get_children().get(programpage__program__readable_id=readable_id)
```